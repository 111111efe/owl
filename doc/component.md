# Component

Components are the reusable, composable widgets. They are designed to be low
level, to be declarative, and with asynchronous rendering.

For example:

```javascript
export class Counter extends Component {
  inlineTemplate = `
    <div t-name="counter">
      <button t-on-click="increment(-1)">-</button>
      <span style="font-weight:bold">Value: <t t-esc="state.counter"/></span>
      <button t-on-click="increment(1)">+</button>
    </div>`;

  state = {
    counter: 0
  };

  constructor(parent, props) {
    super(parent, props);
    this.state.counter = props.initialState || 0;
  }

  increment(delta) {
    this.setState({ counter: this.state.counter + delta });
  }
}
```
