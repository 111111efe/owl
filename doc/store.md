# ðŸ¦‰ Store ðŸ¦‰

## Content

- [Overview](#overview)
- [Example](#example)
- [Reference](#reference)
  - [Public API](#public-api)
  - [Mutations](#mutations)
  - [Actions](#actions)
  - [Getters](#getters)
  - [Connecting a component](#connecting-a-component)

## Overview

Managing the state in an application is not an easy task. In some cases, the
state of an application can be part of the component tree, in a natural way.
However, there are situations where some part of the state need to be displayed
in various parts of the user interface, and then, it is not obvious which
component should own which part of the state.

Owl's solution to this issue is a centralized store. It is a class that owns
some state, and let the developer update it in a structured way (through
mutations and actions). Owl components can then connect to the store, and will
be updated if necessary.

Note: Owl's store is inspired by React Redux and VueX.

## Example

Here is what a simple store look like:

```js
const actions = {
    addTodo({commit}, message) {
        commit('addTodo', message);
    }
};

const mutations = {
    addTodo({state}, message) {
        const todo = {
            id: state.nextId++,
            message,
            isCompleted: false,
        };
        state.todos.push(todo);
    },
};

const state = {
    todos: [],
    nextId: 1,
};

const store = new owl.Store({state, actions, mutations});
store.on('update', () => console.log(store.state));

// updating the state
store.dispatch('addTodo', 'fix all bugs');
```


## Reference

The store is a simple [`owl.EventBus`](event_bus.md) that triggers `update` events whenever its
state is changed.  Note that these events are triggered only after a microtask
tick, so only one event will be triggered for any number of state changes in a
call stack.

Also, it is important to mention that the state is observed (with a `owl.Observer`),
which is the reason why it is able to know if it was changed.  This implies that
state changes need to be done carefully in some cases (adding a new key to an
object, or modifying an array with the `arr[i] = newValue` syntax).  See the
[Observer](observer.md)'s documentation for more details.

### Public API

1. `constructor`
2. `commit`
3. `dispatch`

### Mutations

Mutations are the only way to modify the state.  Changing the state outside a
mutation is not allowed (and should throw an error).  Mutations are synchronous.

### Actions

Actions are used to coordinate state changes.  It is also useful whenever some
asynchronous logic is necessary.  For example, fetching data should be done
in an action.

```js
const actions = {
    async login({commit}) {
        commit('setLoginState', 'pending');
        try {
            const loginInfo = await doSomeRPC('/login/', 'someinfo');
            commit('setLoginState', loginInfo);
        } catch {
            commit('setLoginState', 'error');
        }
    }
};
```

### Getters

Usually, data contained in the store will be stored in a normalized way. For
example,

```js
{
    posts: [{id: 11, authorId: 4, content: 'Greetings'}],
    authors: [{id: 4, name: 'John'}]
}
```

However, the user interface will probably need some denormalized data like

```js
{id: 11, author: {id: 4, name: 'John'}, content: 'Greetings'}
```

This is what `getters` are for: they give a centralized way to process and
transform the data contained in the store.

```js
const getters = {
    getPost({state}, id) {
        const post = state.posts.find(p => p.id === id);
        const author = state.authors.find(a => a.id = post.id);
        return {
            id,
            author,
            content: post.content
        };
    },
};

// somewhere else
const post = store.getters.getPost(id);

```

### Connecting a component

Todo