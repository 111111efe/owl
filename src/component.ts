import h from "../libs/snabbdom/src/h";
import sdAttrs from "../libs/snabbdom/src/modules/attributes";
import sdProps from "../libs/snabbdom/src/modules/props";
import sdListeners from "../libs/snabbdom/src/modules/eventlisteners";
import { init } from "../libs/snabbdom/src/snabbdom";
import { VNode } from "../libs/snabbdom/src/vnode";
import { EventBus } from "./event_bus";
import { QWeb } from "./qweb";
import { idGenerator } from "./utils";

let getId = idGenerator();

//------------------------------------------------------------------------------
// Types/helpers
//------------------------------------------------------------------------------

export interface Env {
  qweb: QWeb;
}

export interface Meta<T extends Env, Props> {
  readonly id: number;
  vnode: VNode | null;
  isStarted: boolean;
  isMounted: boolean;
  isDestroyed: boolean;
  parent: Component<T, any, any> | null;
  children: { [key: number]: Component<T, any, any> };
  // children mapping: from templateID to widgetID
  // should it be a map number => Widget?
  cmap: { [key: number]: number };

  renderId: number;
  renderProps: Props | null;
  renderPromise: Promise<VNode> | null;
  boundHandlers: { [key: number]: any };
}

const patch = init([sdListeners, sdAttrs, sdProps]);

export interface Type<T> extends Function {
  new (...args: any[]): T;
}

//------------------------------------------------------------------------------
// Widget
//------------------------------------------------------------------------------

export class Component<
  T extends Env,
  Props extends {},
  State extends {}
> extends EventBus {
  readonly __owl__: Meta<Env, Props>;
  template: string = "default";
  inlineTemplate: string | null = null;

  get el(): HTMLElement | null {
    return this.__owl__.vnode ? (<any>this).__owl__.vnode.elm : null;
  }

  env: T;
  state: State = <State>{};
  props: Props;
  refs: {
    [key: string]: Component<T, any, any> | HTMLElement | undefined;
  } = {};

  //--------------------------------------------------------------------------
  // Lifecycle
  //--------------------------------------------------------------------------

  /**
   * Creates an instance of Component.
   *
   * The root widget of a component tree needs an environment:
   *
   * ```javascript
   *   const root = new RootWidget(env, props);
   * ```
   *
   * Every other widget simply needs a reference to its parent:
   *
   * ```javascript
   *   const child = new SomeWidget(parent, props);
   * ```
   *
   * Note that most of the time, only the root widget needs to be created by
   * hand.  Other widgets should be created automatically by the framework (with
   * the t-widget directive in a template)
   */
  constructor(parent: Component<T, any, any> | T, props?: Props) {
    super();

    // is this a good idea?
    //   Pro: if props is empty, we can create easily a widget
    //   Con: this is not really safe
    //   Pro: but creating widget (by a template) is always unsafe anyway
    this.props = <Props>props || <Props>{};
    let id: number = getId();
    let p: Component<T, any, any> | null = null;
    if (parent instanceof Component) {
      p = parent;
      this.env = parent.env;
      parent.__owl__.children[id] = this;
    } else {
      this.env = parent;
    }
    this.__owl__ = {
      id: id,
      vnode: null,
      isStarted: false,
      isMounted: false,
      isDestroyed: false,
      parent: p,
      children: {},
      cmap: {},
      renderId: 1,
      renderPromise: null,
      renderProps: props || null,
      boundHandlers: {}
    };
  }

  /**
   * willStart is an asynchronous hook that can be implemented to perform some
   * action before the initial rendering of a component.
   *
   * It will be called exactly once before the initial rendering. It is useful
   * in some cases, for example, to load external assets (such as a JS library)
   * before the widget is rendered.
   *
   * Note that a slow willStart method will slow down the rendering of the user
   * interface.  Therefore, some effort should be made to make this method as
   * fast as possible.
   *
   * Note: this method should not be called manually.
   */
  async willStart() {}

  /**
   * mounted is a hook that is called each time a component is attached to the
   * DOM. This is a good place to add some listeners, or to interact with the
   * DOM, if the component needs to perform some measure for example.
   *
   * Note: this method should not be called manually.
   *
   * @see willUnmount
   */
  mounted() {}

  /**
   * The willUpdateProps is an asynchronous hook, called just before new props
   * are set. This is useful if the component needs some asynchronous task
   * performed, depending on the props (for example, assuming that the props are
   * some record Id, fetching the record data).
   *
   * This hook is not called during the first render (but willStart is called
   * and performs a similar job).
   */
  async willUpdateProps(nextProps: Props) {}

  /**
   * The willPatch hook is called just before the DOM patching process starts.
   * It is not called on the initial render.  This is useful to get some
   * information which are in the DOM.  For example, the current position of the
   * scrollbar
   *
   * Note that at this point, it is not safe to rerender the widget. In
   * particular, updateState calls should be avoided.
   */
  willPatch() {}

  /**
   * This hook is called whenever a component did actually update its props,
   * state or env.
   *
   * This method is not called on the initial render. It is useful to interact
   * with the DOM (for example, through an external library) whenever the
   * component was updated.
   *
   * Updating the widget state in this hook is possible, but not encouraged.
   * One need to be careful, because updates here will cause rerender, which in
   * turn will cause other calls to updated. So, we need to be particularly
   * careful at avoiding endless cycles.
   */
  patched() {}

  /**
   * willUnmount is a hook that is called each time just before a component is
   * unmounted from the DOM. This is a good place to remove some listeners, for
   * example.
   *
   * Note: this method should not be called manually.
   *
   * @see mounted
   */
  willUnmount() {}

  //--------------------------------------------------------------------------
  // Public
  //--------------------------------------------------------------------------

  /**
   * Attach a child widget to a given html element
   *
   * This is most of the time not necessary, since widgets should primarily be
   * created/managed with the t-widget directive in a qweb template.  However,
   * for the cases where we need more control, this method will do what is
   * necessary to make sure all the proper hooks are called (for example,
   * mounted/willUnmount)
   *
   * Note that this method makes a few assumptions:
   * - the child widget is indeed a child of the current widget
   * - the target is inside the dom of the current widget (typically a ref)
   */
  attachChild(child: Component<T, any, any>, target: HTMLElement) {
    target.appendChild(child.el!);
    child.__mount();
  }

  async mount(target: HTMLElement): Promise<void> {
    const vnode = await this._prepare();
    if (this.__owl__.isDestroyed) {
      // widget was destroyed before we get here...
      return;
    }
    this._patch(vnode);
    target.appendChild(this.el!);

    if (document.body.contains(target)) {
      this._visitSubTree(w => {
        if (!w.__owl__.isMounted && this.el!.contains(w.el)) {
          w.__owl__.isMounted = true;
          w.mounted();
          return true;
        }
        return false;
      });
    }
  }

  unmount() {
    if (this.el) {
      this._visitSubTree(w => {
        if (w.__owl__.isMounted) {
          w.willUnmount();
          w.__owl__.isMounted = false;
          return true;
        }
        return false;
      });
      this.el.remove();
    }
  }

  async render(force: boolean = false): Promise<void> {
    if (this.__owl__.isDestroyed) {
      return;
    }
    const renderVDom = this._render(force);
    const renderId = this.__owl__.renderId;
    const vnode = await renderVDom;
    if (renderId === this.__owl__.renderId) {
      // we only update the vnode and the actual DOM if no other rendering
      // occurred between now and when the render method was initially called.
      this._patch(vnode);
    }
  }

  destroy() {
    if (!this.__owl__.isDestroyed) {
      for (let id in this.__owl__.children) {
        this.__owl__.children[id].destroy();
      }
      if (this.__owl__.isMounted) {
        this.willUnmount();
      }
      if (this.el) {
        this.el.remove();
        this.__owl__.isMounted = false;
        delete this.__owl__.vnode;
      }
      if (this.__owl__.parent) {
        let id = this.__owl__.id;
        delete this.__owl__.parent.__owl__.children[id];
        this.__owl__.parent = null;
      }
      this.clear();
      this.__owl__.isDestroyed = true;
    }
  }

  shouldUpdate(nextProps: Props): boolean {
    return true;
  }

  /**
   * This method is the correct way to update the environment of a widget. Doing
   * this will cause a full rerender of the widget and its children, so this is
   * an operation that should not be done frequently.
   *
   * A good usecase for updating the environment would be to update some mostly
   * static config keys, such as a boolean to determine if we are in mobile
   * mode or not.
   */
  async updateEnv(nextEnv: Partial<T>): Promise<void> {
    if (this.__owl__.parent && this.__owl__.parent.env === this.env) {
      this.env = Object.create(this.env);
    }
    Object.assign(this.env, nextEnv);
    if (this.__owl__.isMounted) {
      await this.render(true);
    }
    this.patched();
  }

  async updateProps(
    nextProps: Props,
    forceUpdate: boolean = false
  ): Promise<void> {
    const shouldUpdate = forceUpdate || this.shouldUpdate(nextProps);
    return shouldUpdate ? this._updateProps(nextProps) : Promise.resolve();
  }

  /**
   * This is the safest update method for widget: its job is to update the state
   * and rerender (if widget is mounted).
   *
   * Notes:
   * - it checks if we do not add extra keys to the state.
   * - it is ok to call updateState before the widget is started. In that
   * case, it will simply update the state and will not rerender
   */
  async updateState(nextState: Partial<State>) {
    if (Object.keys(nextState).length === 0) {
      return;
    }
    Object.assign(this.state, nextState);
    if (this.__owl__.isStarted) {
      await this.render();
    }
    this.patched();
  }

  //--------------------------------------------------------------------------
  // Private
  //--------------------------------------------------------------------------

  async _updateProps(nextProps: Props): Promise<void> {
    await this.willUpdateProps(nextProps);
    this.props = nextProps;
    await this.render();
    this.patched();
  }

  _patch(vnode) {
    this.__owl__.renderPromise = null;
    if (this.__owl__.vnode) {
      this.willPatch();
      this.__owl__.vnode = patch(this.__owl__.vnode, vnode);
    } else {
      this.__owl__.vnode = patch(document.createElement(vnode.sel!), vnode);
    }
  }
  async _prepare(): Promise<VNode> {
    this.__owl__.renderProps = this.props;
    this.__owl__.renderPromise = this.willStart().then(() => {
      if (this.__owl__.isDestroyed) {
        return Promise.resolve(h("div"));
      }
      this.__owl__.isStarted = true;
      if (this.inlineTemplate) {
        this.env.qweb.addTemplate(
          this.inlineTemplate,
          this.inlineTemplate,
          true
        );
      }
      return this._render();
    });
    return this.__owl__.renderPromise;
  }

  async _render(force: boolean = false): Promise<VNode> {
    this.__owl__.renderId++;
    const promises: Promise<void>[] = [];
    const template = this.inlineTemplate || this.template;
    let vnode = this.env.qweb.render(template, this, {
      promises,
      handlers: this.__owl__.boundHandlers,
      forceUpdate: force
    });

    // this part is critical for the patching process to be done correctly. The
    // tricky part is that a child widget can be rerendered on its own, which
    // will update its own vnode representation without the knowledge of the
    // parent widget.  With this, we make sure that the parent widget will be
    // able to patch itself properly after
    vnode.key = this.__owl__.id;
    this.__owl__.renderProps = this.props;
    this.__owl__.renderPromise = Promise.all(promises).then(() => vnode);
    return this.__owl__.renderPromise;
  }

  /**
   * Only called by qweb t-widget directive
   */
  _mount(vnode: VNode, elm: HTMLElement): VNode {
    this.__owl__.vnode = patch(elm, vnode);
    this.__mount();
    return this.__owl__.vnode;
  }

  __mount() {
    if (this.__owl__.isMounted) {
      return;
    }
    if (this.__owl__.parent) {
      if (this.__owl__.parent!.__owl__.isMounted) {
        this.__owl__.isMounted = true;
        this.mounted();
        const children = this.__owl__.children;
        for (let id in children) {
          children[id].__mount();
        }
      }
    }
  }

  _visitSubTree(callback: (w: Component<T, any, any>) => boolean) {
    const shouldVisitChildren = callback(this);
    if (shouldVisitChildren) {
      const children = this.__owl__.children;
      for (let id in children) {
        children[id]._visitSubTree(callback);
      }
    }
  }
}

export class PureComponent<E extends Env, P, S> extends Component<E, P, S> {
  shouldUpdate(nextProps: P): boolean {
    for (let k in nextProps) {
      if (nextProps[k] !== this.props[k]) {
        return true;
      }
    }
    return false;
  }
  async updateState(nextState: Partial<S>) {
    for (let k in nextState) {
      if (nextState[k] !== this.state[k]) {
        return super.updateState(nextState);
      }
    }
  }
}
