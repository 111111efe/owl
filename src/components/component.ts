import { FComponent, CComponent, OwlElement, render } from "./core";

export class Component<Props = any, Env = any> {
  static template: string;
  static components: { [key: string]: FComponent | CComponent } = {};
  props: Props;
  env: Env;

  el: HTMLElement | Text | Comment | null = null;
  __owl__: OwlElement | null = null;

  constructor(props: Props, env: Env) {
    this.props = props;
    this.env = env;
  }

  async willStart(): Promise<void> {}

  render(): Promise<void> {
    return render(this.__owl__!);
  }
}
