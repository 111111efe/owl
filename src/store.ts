import { EventBus } from "./event_bus";
import { Component } from "./component";

//------------------------------------------------------------------------------
// Store Definition
//------------------------------------------------------------------------------

interface StoreConfig {
  env?: any;
  state?: any;
  actions?: any;
  mutations?: { [name: string]: any };
}

interface StoreOption {
  debug?: boolean;
}

export class Store extends EventBus {
  state: any;
  actions: any;
  mutations: any;
  _isMutating: boolean = false;
  history: any[] = [];
  debug: boolean;
  env: any;
  observer: Observer;

  constructor(config: StoreConfig, options: StoreOption = {}) {
    super();
    this.debug = options.debug || false;
    this.state = config.state || {};
    this.actions = config.actions;
    this.mutations = config.mutations;
    this.env = config.env;
    this.observer = makeObserver();
    this.observer.allowMutations = false;
    this.observer.observe(this.state);

    if (this.debug) {
      this.history.push({ state: this.state });
    }
  }

  dispatch(action, payload?: any): Promise<void> | void {
    if (!this.actions[action]) {
      throw new Error(`[Error] action ${action} is undefined`);
    }
    const result = this.actions[action](
      {
        commit: this.commit.bind(this),
        dispatch: this.dispatch.bind(this),
        env: this.env,
        state: this.state
      },
      payload
    );
    if (result instanceof Promise) {
      return new Promise((resolve, reject) => {
        result.then(() => resolve());
        result.catch(reject);
      });
    }
  }

  async commit(type, payload?: any) {
    if (!this.mutations[type]) {
      throw new Error(`[Error] mutation ${type} is undefined`);
    }
    const currentRev = this.observer.__rev__;

    this._isMutating = true;
    this.observer.allowMutations = true;
    this.mutations[type].call(
      null,
      { state: this.state, set: this.observer.set },
      payload
    );
    this.observer.allowMutations = false;

    if (this.debug) {
      this.history.push({
        state: this.state,
        mutation: type,
        payload: payload
      });
    }
    await Promise.resolve();
    if (this._isMutating) {
      this._isMutating = false;
      if (currentRev !== this.observer.__rev__) {
        this.trigger("update", this.state);
      }
    }
  }
}

//------------------------------------------------------------------------------
// Observer
//------------------------------------------------------------------------------
interface Observer {
  __rev__: number;
  allowMutations: boolean;
  observe: (val: any) => void;
  set: (target: any, key: number | string, value: any) => void;
}

export function makeObserver(): Observer {
  const observer: Observer = {
    __rev__: 0,
    allowMutations: true,
    observe: observe,
    set: set
  };

  function set(target: any, key: number | string, value: any) {
    addProp(target, key, value);
    target.__rev__++;
    observer.__rev__++;
  }

  function addProp<T extends { __rev__?: number }>(
    obj: T,
    key: string | number,
    value: any
  ) {
    Object.defineProperty(obj, key, {
      enumerable: true,
      get() {
        return value;
      },
      set(newVal) {
        if (!observer.allowMutations) {
          throw new Error(
            `State cannot be changed outside a mutation! (key: "${key}", val: "${newVal}")`
          );
        }
        if (newVal !== value) {
          value = newVal;
          observe(newVal);
          obj.__rev__!++;
          observer.__rev__++;
        }
      }
    });
    observe(value);
  }

  function observeObj<T extends { __rev__?: number }>(obj: T) {
    const keys = Object.keys(obj);
    obj.__rev__ = 0;
    Object.defineProperty(obj, "__rev__", { enumerable: false });
    for (let key of keys) {
      addProp(obj, key, obj[key]);
    }
  }

  const ArrayProto = Array.prototype;
  const ModifiedArrayProto = Object.create(ArrayProto);

  ModifiedArrayProto.push = function(...args) {
    observer.__rev__++;
    this.__rev__++;
    return ArrayProto.push.call(this, ...args);
  };

  ModifiedArrayProto.pop = function(...args) {
    observer.__rev__++;
    this.__rev__++;
    return ArrayProto.pop.call(this, ...args);
  };
  function observeArr(arr: Array<any>) {
    (<any>arr).__rev__ = 0;
    Object.defineProperty(arr, "__rev__", { enumerable: false });
    (<any>arr).__proto__ = ModifiedArrayProto;
    for (let i = 0; i < arr.length; i++) {
      observe(arr[i]);
    }
  }

  function observe(value: any) {
    if (typeof value !== "object") {
      return;
    }
    if ("__rev__" in value) {
      // already observed
      return;
    }
    if (Array.isArray(value)) {
      observeArr(value);
    } else {
      observeObj(value);
    }
  }

  return observer;
}

//------------------------------------------------------------------------------
// Connect function
//------------------------------------------------------------------------------

function setStoreProps(__widget__: any, storeProps: any) {
  __widget__.currentStoreProps = storeProps;
  __widget__.currentStoreRevs = {};
  __widget__.currentStoreRev = storeProps.__rev__;
  for (let key in storeProps) {
    __widget__.currentStoreRevs[key] = storeProps[key].__rev__;
  }
}

export function connect(mapStateToProps) {
  return function(Comp) {
    return class extends Comp {
      constructor(parent, props?: any) {
        const env = parent instanceof Component ? parent.env : parent;
        const ownProps = Object.assign({}, props || {});
        const storeProps = mapStateToProps(env.store.state, ownProps);
        const mergedProps = Object.assign({}, props || {}, storeProps);
        super(parent, mergedProps);
        setStoreProps(this.__widget__, storeProps);
        this.__widget__.ownProps = ownProps;
      }
      mounted() {
        this.env.store.on("update", this, () => {
          const ownProps = this.__widget__.ownProps;
          const storeProps = mapStateToProps(this.env.store.state, ownProps);
          let didChange = false;
          if (this.__widget__.currentStoreRev !== storeProps.__rev__) {
            setStoreProps(this.__widget__, storeProps);
            didChange = true;
          } else {
            const revs = this.__widget__.currentStoreRevs;
            for (let key in storeProps) {
              const val = storeProps[key];
              if (val.__rev__ !== revs[key]) {
                didChange = true;
                revs[key] = val.__rev__;
                this.__widget__.currentStoreProps[key] = val;
              }
            }
          }
          if (didChange) {
            this.updateProps(ownProps, false);
          }
        });
        super.mounted();
      }
      willUnmount() {
        this.env.store.off("update", this);
        super.willUnmount();
      }
      updateProps(nextProps, forceUpdate) {
        if (this.__widget__.ownProps !== nextProps) {
          this.__widget__.currentStoreProps = mapStateToProps(
            this.env.store.state,
            nextProps
          );
        }
        this.__widget__.ownProps = nextProps;
        const mergedProps = Object.assign(
          {},
          nextProps,
          this.__widget__.currentStoreProps
        );
        return super.updateProps(mergedProps, forceUpdate);
      }
    };
  };
}
