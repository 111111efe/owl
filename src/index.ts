import { QWeb } from "./qweb";
import { EventBus } from "./event_bus";
import { Component, PureComponent } from "./component";
import { Store, connect } from "./store";
import * as utils from "./utils";
export const core = {
  QWeb,
  EventBus,
  Component,
  PureComponent,
  Store,
  connect,
  utils
};
