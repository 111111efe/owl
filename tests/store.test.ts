import { Store, connect } from "../src/store";
import { Component, Env } from "../src/component";
import {
  nextMicroTick,
  makeTestWEnv,
  makeTestFixture,
  nextTick
} from "./helpers";

describe("basic use", () => {
  test("commit a mutation", () => {
    const state = { n: 1 };
    const mutations = {
      inc(state, delta) {
        state.n += delta;
      }
    };
    const store = new Store({ state, mutations });

    expect(store.state.n).toBe(1);
    store.commit("inc", 14);
    expect(store.state.n).toBe(15);
  });

  test("dispatch an action", () => {
    const state = { n: 1 };
    const mutations = {
      inc(state, delta) {
        state.n += delta;
      }
    };
    const actions = {
      inc({ commit }, delta) {
        commit("inc", delta);
      }
    };
    const store = new Store({ state, mutations, actions });

    expect(store.state.n).toBe(1);
    store.dispatch("inc", 14);
    expect(store.state.n).toBe(15);
  });

  test("env is given to actions", () => {
    expect.assertions(1);
    const someEnv = {};
    const actions = {
      someaction({ env }) {
        expect(env).toBe(someEnv);
      }
    };
    const store = new Store({ state: {}, actions, env: someEnv });

    store.dispatch("someaction");
  });

  test("multiple commits trigger one update", async () => {
    let updateCounter = 0;
    const state = { n: 1 };
    const mutations = {
      inc(state, delta) {
        state.n += delta;
      }
    };
    const store = new Store({ state, mutations });
    store.on("update", null, () => updateCounter++);

    store.commit("inc", 14);
    expect(updateCounter).toBe(0);
    store.commit("inc", 50);
    expect(updateCounter).toBe(0);
    await nextMicroTick();
    expect(updateCounter).toBe(1);
  });
});

describe("connecting a component to store", () => {
  let fixture: HTMLElement;
  let env: Env;

  beforeEach(() => {
    fixture = makeTestFixture();
    env = makeTestWEnv();
  });

  afterEach(() => {
    fixture.remove();
  });

  class App extends Component<any, any, any> {
    inlineTemplate = `
            <div>
                <t t-foreach="props.todos" t-as="todo">
                    <t t-widget="Todo" t-props="todo"/>
                </t>
            </div>`;
    widgets = { Todo };
  }
  class Todo extends Component<any, any, any> {
    inlineTemplate = `<span><t t-esc="props.msg"/></span>`;
  }

  test("connecting a component works", async () => {
    const state = { todos: [] };
    const mutations = {
      addTodo(state, msg) {
        state.todos.push({ msg });
      }
    };
    const TodoApp = connect(s => s)(App);
    const store = new Store({ state, mutations });
    (<any>env).store = store;
    const app = new TodoApp(env);

    await app.mount(fixture);
    expect(fixture.innerHTML).toMatchSnapshot();

    store.commit("addTodo", "hello");
    await nextTick();
    expect(fixture.innerHTML).toMatchSnapshot();
  });

  test("connected child components with custom hooks", async () => {
    let steps: any = [];
    class Child extends Component<any, any, any> {
      inlineTemplate = `<div/>`;
      mounted() {
        steps.push("child:mounted");
      }
      willUnmount() {
        steps.push("child:willUnmount");
      }
      destroyed() {
        steps.push("child:destroyed");
      }
    }

    const ConnectedChild = connect(s => s)(Child);

    class Parent extends Component<any, any, any> {
      inlineTemplate = `
        <div>
            <t t-if="state.child" t-widget="ConnectedChild"/>
        </div>`;
      widgets = { ConnectedChild };

      constructor(env: Env) {
        super(env);
        this.state = { child: true };
      }
    }

    const store = new Store({ state: {} });
    (<any>env).store = store;
    const parent = new Parent(env);

    await parent.mount(fixture);
    expect(steps).toEqual(["child:mounted"]);

    await parent.updateState({ child: false });
    expect(steps).toEqual([
      "child:mounted",
      "child:willUnmount",
      "child:destroyed"
    ]);
  });

  test("connect receives ownprops as second argument", async () => {
    const state = { todos: [{ id: 1, text: "jupiler" }] };
    let nextId = 2;
    const mutations = {
      addTodo(state, text) {
        state.todos.push({ text, id: nextId++ });
      }
    };
    const store = new Store({ state, mutations });

    class TodoItem extends Component<any, any, any> {
      inlineTemplate = `<span><t t-esc="props.text"/></span>`;
    }
    const ConnectedTodo = connect((state, props) => {
      const todo = state.todos.find(t => t.id === props.id);
      return todo;
    })(TodoItem);

    class TodoList extends Component<any, any, any> {
      inlineTemplate = `<div>
          <t t-foreach="props.todos" t-as="todo">
            <t t-widget="ConnectedTodo" t-props="todo"/>
          </t>
        </div>`;
      widgets = { ConnectedTodo };
    }

    const ConnectedTodoList = connect(state => state)(TodoList);

    (<any>env).store = store;
    const app = new ConnectedTodoList(env);

    await app.mount(fixture);
    expect(fixture.innerHTML).toBe("<div><span>jupiler</span></div>");

    store.commit("addTodo", "hoegaarden");
    await nextTick();
    expect(fixture.innerHTML).toBe(
      "<div><span>jupiler</span><span>hoegaarden</span></div>"
    );
  });

  test("connected component is updated when props are updated", async () => {
    class Beer extends Component<any, any, any> {
      inlineTemplate = `<span><t t-esc="props.name"/></span>`;
    }
    const ConnectedBeer = connect((state, props) => {
      return state.beers[props.id];
    })(Beer);

    class App extends Component<any, any, any> {
      inlineTemplate = `<div>
            <t t-widget="ConnectedBeer" t-props="{id: state.beerId}"/>
        </div>`;
      widgets = { ConnectedBeer };
      state = { beerId: 1 };
    }

    const state = { beers: { 1: { name: "jupiler" }, 2: { name: "kwak" } } };
    const store = new Store({ state });
    (<any>env).store = store;
    const app = new App(env);

    await app.mount(fixture);
    expect(fixture.innerHTML).toBe("<div><span>jupiler</span></div>");

    await app.updateState({ beerId: 2 });
    expect(fixture.innerHTML).toBe("<div><span>kwak</span></div>");
  });
});
