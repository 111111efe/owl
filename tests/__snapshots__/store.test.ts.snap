// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connecting a component to store connecting a component works 1`] = `"<div></div>"`;

exports[`connecting a component to store connecting a component works 2`] = `"<div><span>hello</span></div>"`;
